diff --git "a/clang/lib/Sema/SemaExpr.cpp" "b/clang/lib/Sema/SemaExpr.cpp"
index 736e76152fe4..957e8070a58f 100644
--- "a/clang/lib/Sema/SemaExpr.cpp"
+++ "b/clang/lib/Sema/SemaExpr.cpp"
@@ -12185,12 +12185,11 @@ QualType Sema::CheckCompareOperands(ExprResult &LHS, ExprResult &RHS,
       // since users tend to want to compare addresses.
     } else if ((LHSIsNull && LHSType->isIntegerType()) ||
                (RHSIsNull && RHSType->isIntegerType())) {
-      if (IsOrdered) {
-        isError = getLangOpts().CPlusPlus;
-        DiagID =
-          isError ? diag::err_typecheck_ordered_comparison_of_pointer_and_zero
-                  : diag::ext_typecheck_ordered_comparison_of_pointer_and_zero;
-      }
+        // [MSVC Compatibility]
+        //isError = getLangOpts().CPlusPlus;
+        //DiagID =
+        //  isError ? diag::err_typecheck_ordered_comparison_of_pointer_and_zero
+        //          : diag::ext_typecheck_ordered_comparison_of_pointer_and_zero;
     } else if (getLangOpts().CPlusPlus) {
       DiagID = diag::err_typecheck_comparison_of_pointer_integer;
       isError = true;
