diff --git aclanglibASTASTContext.cpp bclanglibASTASTContext.cpp
index ac6f98e91f75..44914bf973b3 100644
--- aclanglibASTASTContext.cpp
+++ bclanglibASTASTContext.cpp
@@ -10430,6 +10430,8 @@ QualType ASTContextgetCorrespondingSignedType(QualType T) const {
     return SatFractTy;
   case BuiltinTypeSatULongFract
     return SatLongFractTy;
+  case BuiltinTypeBool [MSVC Compatibility]
+    return BoolTy;
   default
     llvm_unreachable(Unexpected unsigned integer or fixed point type);
   }
