diff --git a/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp b/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp
index 5d00fa56e888..b373e1b3b849 100644
--- a/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp
+++ b/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp
@@ -120,7 +120,7 @@ bool DisableLIRP::All;
 static cl::opt<bool, true>
     DisableLIRPAll("disable-" DEBUG_TYPE "-all",
                    cl::desc("Options to disable Loop Idiom Recognize Pass."),
-                   cl::location(DisableLIRP::All), cl::init(false),
+                   cl::location(DisableLIRP::All), cl::init(true),// [MSVC Compability]
                    cl::ReallyHidden);
 
 bool DisableLIRP::Memset;
@@ -128,7 +128,7 @@ static cl::opt<bool, true>
     DisableLIRPMemset("disable-" DEBUG_TYPE "-memset",
                       cl::desc("Proceed with loop idiom recognize pass, but do "
                                "not convert loop(s) to memset."),
-                      cl::location(DisableLIRP::Memset), cl::init(false),
+                      cl::location(DisableLIRP::Memset), cl::init(true),// [MSVC Compability]
                       cl::ReallyHidden);
 
 bool DisableLIRP::Memcpy;
