diff --git a/clang/lib/Lex/TokenLexer.cpp b/clang/lib/Lex/TokenLexer.cpp
index 8fcdbf4dbcb7..fa0a781337e7 100644
--- a/clang/lib/Lex/TokenLexer.cpp
+++ b/clang/lib/Lex/TokenLexer.cpp
@@ -815,7 +815,20 @@ bool TokenLexer::pasteTokens(Token &LHSTok, ArrayRef<Token> TokenStream,
 
     // Lex the resultant pasted token into Result.
     Token Result;
-
+    // [MSVC Compatibility]
+    if ((LHSTok.is(tok::period) || LHSTok.is(tok::arrow)) &&
+        RHS.isAnyIdentifier()) {
+      // Handle '.##identifier' or '->##identifier'
+      // Returns true the caller should immediately return the token.
+      return true;
+    }
     if (LHSTok.isAnyIdentifier() && RHS.isAnyIdentifier()) {
       // Common paste case: identifier+identifier = identifier.  Avoid creating
       // a lexer and other overhead.
